### PR TITLE
changes in the computation of the metal_dmat 

### PR DESCRIPTION
**Pull request by vserret**
The metal matrix should depend on : 
- zref 
- alpha_abs1 and alpha_abs2 (which can be different than alpha_met) 

I've also rewrite z1_abs2, z2_abs1 ... at the end to avoid unuseful computation in the case of the cross correlation.

### CODE REVIEW

**Comment by ngbusca** [comment]
@vserret thanks for this. You will have to regenerate the reference files for the tests to pass. Please also take a look at the coadacy style suggestions.

File: bin/do_metal_dmat.py [hunk 1]
```diff
@@ -106,7 +106,6 @@ def calc_metal_dmat(abs_igm1,abs_igm2,p):
 
     if args.nproc is None:
         args.nproc = cpu_count()//2
-
```
Comment (londumas) [inline bin/do_metal_dmat.py]: Could you restore this line, no need to change this

File: bin/do_metal_dmat.py [hunk 2]
```diff
@@ -133,7 +132,7 @@ def calc_metal_dmat(abs_igm1,abs_igm2,p):
     lambda_abs  = constants.absorber_IGM[args.lambda_abs]
     if args.lambda_abs2: lambda_abs2 = constants.absorber_IGM[args.lambda_abs2]
     else: lambda_abs2 = constants.absorber_IGM[args.lambda_abs]
-
+
```
Comment (londumas) [inline bin/do_metal_dmat.py]: Could you remove trailing spaces, it is a recomendation of Codacy

File: bin/do_metal_dmat.py [hunk 3]
```diff
@@ -225,7 +224,7 @@ def calc_metal_dmat(abs_igm1,abs_igm2,p):
     cf.npix = len(data)
     cf.data = data
     cf.ndata = ndata
-    cf.alpha_met = args.metal_alpha
+    #cf.alpha_met = args.metal_alpha
```
Comment (londumas) [inline bin/do_metal_dmat.py]: why keeping this?

File: bin/do_metal_dmat.py [hunk 4]
```diff
@@ -258,7 +257,7 @@ def calc_metal_dmat(abs_igm1,abs_igm2,p):
 
     abs_igm = [args.lambda_abs]+args.abs_igm
     print("abs_igm = {}".format(abs_igm))
-
+
```
Comment (londumas) [inline bin/do_metal_dmat.py]: same for trailing spaces

File: bin/do_metal_dmat.py [hunk 5]
```diff
@@ -270,9 +269,16 @@ def calc_metal_dmat(abs_igm1,abs_igm2,p):
         abs_igm_2 = copy.deepcopy(abs_igm)
     else:
         abs_igm_2 = [lambda_abs2_type]
-
```
Comment (londumas) [inline bin/do_metal_dmat.py]: please, keep lines as they are

File: py/picca/cf.py [hunk 6]
```diff
@@ -266,15 +266,14 @@ def fill_dmat(l1,l2,r1,r2,w1,w2,ang,wdm,dm,same_half_plate,order1,order2):
             dm[bb+np*nt*ba] += we[k]*(eta5[bb]+eta6[bb]*dl2[j]+eta7[bb]*dl1[i]+eta8[bb]*dl1[i]*dl2[j])\
              - we[k]*(eta1[i+n1*bb]+eta3[i+n1*bb]*dl2[j]+eta2[j+n2*bb]+eta4[j+n2*bb]*dl1[i])
 
-def metal_dmat(pix,abs_igm1="LYA",abs_igm2="SiIII(1207)"):
-
```
Comment (londumas) [inline py/picca/cf.py]: why remove this space?

File: py/picca/cf.py [hunk 7]
```diff
@@ -266,15 +266,14 @@ def fill_dmat(l1,l2,r1,r2,w1,w2,ang,wdm,dm,same_half_plate,order1,order2):
...
+def metal_dmat(pix,abs_igm1="LYA",abs_igm2="SiIII(1207)"):    
     dm = sp.zeros(np*nt*ntm*npm)
     wdm = sp.zeros(np*nt)
     rpeff = sp.zeros(ntm*npm)
     rteff = sp.zeros(ntm*npm)
     zeff = sp.zeros(ntm*npm)
     weff = sp.zeros(ntm*npm)
-
+
```
Comment (londumas) [inline py/picca/cf.py]: same for this trailing space

File: py/picca/cf.py [hunk 8]
```diff
@@ -298,11 +295,8 @@ def metal_dmat(pix,abs_igm1="LYA",abs_igm2="SiIII(1207)"):
                         ( (d1.fid<=500 and d2.fid<=500) or (d1.fid>500 and d2.fid>500) )
                 ang = d1^d2
                 r2 = d2.r_comov
-                z2_abs1 = 10**d2.ll/constants.absorber_IGM[abs_igm1]-1
-                r2_abs1 = cosmo.r_comoving(z2_abs1)
                 z2_abs2 = 10**d2.ll/constants.absorber_IGM[abs_igm2]-1
                 r2_abs2 = cosmo.r_comoving(z2_abs2)
-
```
Comment (londumas) [inline py/picca/cf.py]: same for this space

File: py/picca/cf.py [hunk 9]
```diff
@@ -356,13 +350,17 @@ def metal_dmat(pix,abs_igm1="LYA",abs_igm2="SiIII(1207)"):
 
 
                 if (not(x_correlation) and (abs_igm1 != abs_igm2)) or (x_correlation and (lambda_abs == lambda_abs2)):
-
```
Comment (londumas) [inline py/picca/cf.py]: same here

**Comment by londumas** [comment]
@vserret, I made a review.
I don't understand why do we have to remake the metal_dmat files.
What are the values that change?
Should we apply these correction to do_metal_xdmat?

Thank you for the work.

**Comment by vserret** [comment]
In master branch : 
l.338 :  zwe12 = (1+z1_abs1[:,None])**(alpha_met-1)*(1+z2_abs2)**(alpha_met-1)/(3.25)**(2*alpha_met-2)

First, if you compute the metal matrix for Lya - SiII, you use alpha = alpha_met for igm_abs1 = Lya and it should be args.z-evol. 
Secondly, since we have an option --zref we should have (1+zref) instead of 3.25(=1+2.25) 
The values of zwe12 will change for Lya-metal metal matrix, not for the others. 

These changes will not appear in the cross with qso since you use : 
wdq = wd*wq which does not depend on the redshift and on the alphas

**Comment by londumas** [comment]
@vserret, thanks for the answer.
However it seems to me that the issue is also in the cross at line 287 of xcf.py it shows that the weight assigned to the pixel is actually the one assigned to Lya.

**Comment by vserret** [comment]
Do you want me to correct it or you want to create a new issue for the cross ?

**Comment by londumas** [comment]
@vserret, let's create another issue for the cross. Thank you for the answer.

**Comment by vserret** [comment]
@ngbusca ready to be merge

**Comment by ngbusca** [comment]
@vserret can you post the results of fits of data with master and your branch?

**Comment by vserret** [comment]
@ngbusca the results for the lya auto fit : 
with master : 
ap = 1.0411 +/- 0.0398
at = 0.9858 +/- 0.0474
beta_LYA = 1.2777 +/- 0.0676
bias_LYA = -0.1846 +/- 0.0047
bias_SiII(1190) = -0.0060 +/- 0.0012
bias_SiII(1193) = 0.0012 +/- 0.0027
bias_SiII(1260) = -0.0029 +/- 0.0015
bias_SiIII(1207) = -0.0072 +/- 0.0019
bias_gamma = 0.0736 +/- 0.0623
beta_hcd = 2.5499 +/- 0.7737
L0_hcd = 14.4974 +/- 0.4092
bias_hcd = -0.0187 +/- 0.0037


with this branch : 
ap = 1.0411 +/- 0.0398
at = 0.9858 +/- 0.0475
beta_LYA = 1.2773 +/- 0.0680
bias_LYA = -0.1846 +/- 0.0048
bias_SiII(1190) = -0.0060 +/- 0.0012
bias_SiII(1193) = 0.0012 +/- 0.0028
bias_SiII(1260) = -0.0029 +/- 0.0015
bias_SiIII(1207) = -0.0072 +/- 0.0020
bias_gamma = 0.0736 +/- 0.0624
beta_hcd = 2.5548 +/- 0.7828
L0_hcd = 14.4970 +/- 0.4183
bias_hcd = -0.0187 +/- 0.0038

**Pull request closed without merging**